### PR TITLE
Node.js bindings for `SlotCommitment`

### PR DESCRIPTION
**Pull request by abdulmth**
# Description of change

Add `SlotCommitment` and it's related types in node.js

related to https://github.com/iotaledger/iota-sdk/pull/642

### CODE REVIEW

File: sdk/src/types/block/slot/commitment.rs [hunk 1]
```diff
@@ -19,7 +19,7 @@ pub struct SlotCommitment {
     /// It is calculated based on genesis timestamp and the duration of a slot.
     index: SlotIndex,
     /// The commitment ID of the previous slot.
-    #[cfg_attr(feature = "serde", serde(rename = "prevId"))]
+    #[serde(rename = "prevId")]
```
Comment (abdulmth) [inline sdk/src/types/block/slot/commitment.rs:25]: `SlotCommitment` already has `serde` feature.

File: bindings/nodejs/lib/types/utils/bridge/utils.ts [hunk 2]
```diff
@@ -187,3 +194,15 @@ export interface __OutputIdToUtxoInput__ {
         outputId: string;
     };
 }
+
+// Modified `SlotCommitment` with bigint types converted to strings.
+type SlotCommitmentConverted = Omit<
```
Comment (abdulmth) [inline bindings/nodejs/lib/types/utils/bridge/utils.ts:193]: Could be an overkill as we can use `any` but it doesn't harm imo.

File: sdk/src/types/block/slot/commitment.rs [hunk 3]
```diff
@@ -78,3 +78,65 @@ impl SlotCommitment {
...
+                dto.index,
+                dto.previous_slot_commitment_id,
+                dto.roots_id,
+                dto.cumulative_weight,
+            )
+        }
+    }
+}
+
+#[cfg(test)]
```
Comment (abdulmth) [inline sdk/src/types/block/slot/commitment.rs:86]: We could move it to `tests/` if needed, but I prefer it here.

File: bindings/nodejs/tests/client/utilityMethods.spec.ts [hunk 4]
```diff
@@ -130,4 +131,17 @@ describe('Client utility methods', () => {
         const validSignature = Utils.verifyEd25519Signature(signature, message);
         expect(validSignature).toBeTruthy();
     });
+
+    it('slot commitment id', async () => {
+        let slotCommitment = new SlotCommitment(
```
Comment (abdulmth) [inline bindings/nodejs/tests/client/utilityMethods.spec.ts:136]: Test values are arbitrary, just to test consistency with Rust.

File: bindings/nodejs/lib/types/utils/type_aliases.ts [hunk 5]
```diff
@@ -2,3 +2,4 @@
 // SPDX-License-Identifier: Apache-2.0
 
 export type u16 = number;
+export type u64 = bigint;
```
Comment (abdulmth) [inline bindings/nodejs/lib/types/utils/type_aliases.ts:5]: JS's `Number.MAX_SAFE_INTEGER` < Rust's `u64::MAX`  so `u64`  should be `bigint` in JS.

File: bindings/nodejs/lib/utils/utils.ts [hunk 6]
```diff
@@ -324,4 +325,25 @@ export class Utils {
             data: { mnemonic },
         });
     }
+
+    /**
+     * Derives the `SlotCommitmentId` of the `SlotCommitment`.
+     */
+    static computeSlotCommitmentId(
```
Comment (abdulmth) [inline bindings/nodejs/lib/utils/utils.ts:332]: using a util function so we don't need to have bindings method calls in `types/`

File: bindings/core/src/method_handler/utils.rs [hunk 7]
```diff
@@ -103,6 +104,10 @@ pub(crate) fn call_utils_method_internal(method: UtilsMethod) -> Result<Response
             Response::Bool(public_key.verify_keccak256(&signature, &message))
         }
         UtilsMethod::OutputIdToUtxoInput { output_id } => Response::Input(UtxoInput::from(output_id)),
+        UtilsMethod::ComputeSlotCommitmentId { dto } => {
+            let slot_commitment_id: SlotCommitmentId = SlotCommitment::from(dto).id();
+            Response::SlotCommitmentId(slot_commitment_id)
```
Comment (kwek20) [inline bindings/core/src/method_handler/utils.rs]: Minify this please 🙏

File: bindings/nodejs/lib/types/block/slot/commitmemt.ts
Comment (kwek20) [inline bindings/nodejs/lib/types/block/slot/commitmemt.ts:1]: typo in the file name

File: bindings/nodejs/lib/types/utils/type_aliases.ts (hunk 5 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/utils/type_aliases.ts:5]: yea checks out, we should update all other bigints to this as well

File: sdk/src/types/block/slot/commitment.rs (hunk 3 above)
Comment (kwek20) [reply sdk/src/types/block/slot/commitment.rs:86]: pub(crate) tests should always be in the file, so this is perfect

**Review by kwek20 — DISMISSED** [review]

File: bindings/nodejs/lib/types/utils/type_aliases.ts (hunk 5 above)
Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/utils/type_aliases.ts:5]: Can we create an issue then?

Comment (kwek20) [reply bindings/nodejs/lib/types/utils/type_aliases.ts:5]: https://github.com/iotaledger/iota-sdk/issues/966 added here

File: bindings/core/src/method/utils.rs [hunk 8]
```diff
@@ -148,4 +149,8 @@ pub enum UtilsMethod {
     /// Creates a UTXOInput from outputId.
     #[serde(rename_all = "camelCase")]
     OutputIdToUtxoInput { output_id: OutputId },
+
```
Comment (Thoralf-M) [inline bindings/core/src/method/utils.rs]:
```suggestion
```

File: bindings/core/src/method_handler/utils.rs [hunk 9]
```diff
@@ -10,6 +10,7 @@ use iota_sdk::{
             input::UtxoInput,
             output::{AccountId, FoundryId, InputsCommitment, NftId, Output, OutputId, Rent, TokenId},
             payload::{transaction::TransactionEssence, TransactionPayload},
+            slot::{SlotCommitment, SlotCommitmentId},
```
Comment (Thoralf-M) [inline bindings/core/src/method_handler/utils.rs]: Unused import SlotCommitmentId

File: bindings/core/src/method/utils.rs [hunk 10]
```diff
@@ -148,4 +149,7 @@ pub enum UtilsMethod {
     /// Creates a UTXOInput from outputId.
     #[serde(rename_all = "camelCase")]
     OutputIdToUtxoInput { output_id: OutputId },
+    /// Computes the slot commitment id from a slot commitment.
+    #[serde(rename_all = "camelCase")]
```
Comment (kwek20) [inline bindings/core/src/method/utils.rs:157]: rename not needed actually

Comment (abdulmth) [reply bindings/core/src/method/utils.rs:157]: Done

**Review by kwek20 — DISMISSED** [review]

File: bindings/nodejs/lib/types/utils/bridge/utils.ts [hunk 11]
```diff
@@ -187,3 +188,15 @@ export interface __OutputIdToUtxoInput__ {
...
+    SlotCommitment,
+    'index' | 'cumulativeWeight'
+> & { index: string; cumulativeWeight: string };
+export interface __ComputeSlotCommitmentId__ {
+    name: 'computeSlotCommitmentId';
+    data: {
+        dto: SlotCommitmentConverted;
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/utils/bridge/utils.ts]: Bit weird to have `dto` as field name

**Review by Thoralf-M — DISMISSED** [review]

**Review by kwek20 — DISMISSED** [review]

File: bindings/core/src/method_handler/utils.rs [hunk 12]
```diff
@@ -10,6 +10,7 @@ use iota_sdk::{
             input::UtxoInput,
             output::{AccountId, FoundryId, InputsCommitment, NftId, Output, OutputId, Rent, TokenId},
             payload::{transaction::TransactionEssence, TransactionPayload},
+            slot::SlotCommitment,
             protocol::calc_slot_index,
```
Comment (Thoralf-M) [inline bindings/core/src/method_handler/utils.rs]:
```suggestion
            protocol::calc_slot_index,
            slot::SlotCommitment,
```

**Review by Thoralf-M — DISMISSED** [review]

**Review by kwek20 — DISMISSED** [review]

File: sdk/src/types/block/slot/commitment.rs [hunk 13]
```diff
@@ -78,3 +78,67 @@ impl SlotCommitment {
         SlotCommitmentId::from(bytes)
     }
 }
+
+pub(crate) mod dto {
```
Comment (DaughterOfMars) [inline sdk/src/types/block/slot/commitment.rs]: But we don't need a Dto at all?

**Review by DaughterOfMars — CHANGES_REQUESTED** [review]
Remove `SlotCommitmentDto`

File: bindings/nodejs/lib/types/block/slot/commitment.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+class SlotCommitment {
+    /**
+     * The slot index of this commitment.
+     * It is calculated based on genesis timestamp and the duration of a slot.
+     */
+    index: SlotIndex;
+    /**
+     * The commitment ID of the previous slot.
+     */
+    previousSlotCommitmentId: SlotCommitmentId;
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/block/slot/commitment.ts]:
```suggestion
    prevId: SlotCommitmentId;
```

**Review by kwek20 — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/core/src/method_handler/utils.rs [hunk 15]
```diff
@@ -112,6 +113,9 @@ pub(crate) fn call_utils_method_internal(method: UtilsMethod) -> Result<Response
             Response::Bool(public_key.verify_keccak256(&signature, &message))
         }
         UtilsMethod::OutputIdToUtxoInput { output_id } => Response::Input(UtxoInput::from(output_id)),
+        UtilsMethod::ComputeSlotCommitmentId { slot_commitment } => {
+            Response::SlotCommitmentId(SlotCommitment::from(slot_commitment).id())
```
Comment (DaughterOfMars) [inline bindings/core/src/method_handler/utils.rs]: Conversion is no longer needed

File: bindings/core/src/method_handler/utils.rs [hunk 16]
```diff
@@ -11,6 +11,7 @@ use iota_sdk::{
             output::{AccountId, FoundryId, InputsCommitment, NftId, Output, OutputId, Rent, TokenId},
             payload::{transaction::TransactionEssence, TransactionPayload},
             protocol::calc_slot_index,
+            slot::SlotCommitment,
```
Comment (DaughterOfMars) [inline bindings/core/src/method_handler/utils.rs]: No longer used

**Review by DaughterOfMars — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Pull request merged**